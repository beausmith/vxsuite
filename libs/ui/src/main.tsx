import styled from 'styled-components';

interface Props {
  padded?: boolean;
  noOverflow?: boolean;
}

export const Main = styled('main')<Props>`
  display: flex;
  overflow: ${({ noOverflow }) => (noOverflow ? undefined : 'auto')};
  padding: ${({ padded }) => (padded ? '1rem 0.5rem 2rem' : undefined)};
`;

interface ChildProps {
  center?: boolean;
  flexContainer?: boolean;
  flexDirection?: React.CSSProperties['flexDirection'];
  maxWidth?: boolean;
  narrow?: boolean;
}

export const MainChild = styled('div')<ChildProps>`
  display: ${({ flexContainer }) => (flexContainer ? 'flex' : undefined)};
  flex: ${({ flexContainer }) => (flexContainer ? 1 : undefined)};
  flex-direction: ${({ flexDirection = 'column' }) => flexDirection};
  margin: ${({ center = false }) => (center ? 'auto' : undefined)};
  max-width: ${({ maxWidth = true, narrow = false }) =>
    narrow ? '50%' : maxWidth ? '35rem' : undefined};
`;

export const MainChildFlexRow = styled.div`
  display: flex;
  flex: 1;
`;
