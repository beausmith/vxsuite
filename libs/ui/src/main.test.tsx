import React from 'react';
import { render } from '@testing-library/react';
import 'jest-styled-components';

import { Main, MainChild, MainChildFlexRow } from './main';

describe('renders Main', () => {
  test('with defaults', () => {
    const { container, getByText } = render(
      <Main>
        <MainChild>MainChild</MainChild>
      </Main>
    );
    const main = container.firstChild;
    expect(main).toHaveStyleRule('overflow', 'auto');
    expect(main).not.toHaveStyleRule('padding');

    const child = getByText('MainChild');
    expect(child).not.toHaveStyleRule('display');
    expect(child).not.toHaveStyleRule('flex');
    expect(child).toHaveStyleRule('flex-direction', 'column');
    expect(child).toHaveStyleRule('max-width', '35rem');
  });

  test('with all non-default Main options and common MainChild options', () => {
    const { container, getByText } = render(
      <Main noOverflow padded>
        <MainChild center narrow>
          MainChild
        </MainChild>
      </Main>
    );
    const main = container.firstChild;
    expect(main).not.toHaveStyleRule('overflow');
    expect(main).toHaveStyleRule('padding', '1rem 0.5rem 2rem');

    const child = getByText('MainChild');
    expect(child).toHaveStyleRule('margin', 'auto');
    expect(child).toHaveStyleRule('max-width', '50%');
  });

  test('with MainChild with no max width', () => {
    const { getByText } = render(
      <Main>
        <MainChild flexContainer maxWidth={false}>
          MainChild
        </MainChild>
      </Main>
    );
    const child = getByText('MainChild');
    expect(child).toHaveStyleRule('display', 'flex');
    expect(child).toHaveStyleRule('flex', '1');
    expect(child).not.toHaveStyleRule('max-width');
  });

  test('with columns in the main child', () => {
    const { getByText } = render(
      <Main>
        <MainChildFlexRow>MainChildFlexRow</MainChildFlexRow>
      </Main>
    );
    expect(getByText('MainChildFlexRow')).not.toHaveStyleRule('flex-direction');
  });
});
