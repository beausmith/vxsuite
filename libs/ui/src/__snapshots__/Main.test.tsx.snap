// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Main with MainChild with no max width 1`] = `
<main
  class="sc-bdvvaa lyEiN"
>
  <div
    class="sc-gsDJrp jxSnOT"
  >
    foo
  </div>
</main>
`;

exports[`renders Main with all non-default Main options and common MainChild options 1`] = `
<main
  class="sc-bdvvaa hwUMQP"
>
  <div
    class="sc-gsDJrp gTPMtF"
  >
    foo
  </div>
</main>
`;

exports[`renders Main with defaults 1`] = `
<main
  class="sc-bdvvaa lyEiN"
>
  <div
    class="sc-gsDJrp hkJJbd"
  >
    foo
  </div>
</main>
`;
